### ISSUE TITLE
Maxed out skills resets while testing map

### ISSUE THREAD

**Issue by sh4z4n** [multitheftauto/mtasa-resources#131]
**Which resource(s) have this problem?**
defaultstats

**To Reproduce**
Steps to reproduce the behaviour:
1. Start Freeroam, then defaultstats.
2. Spawn any bicycle, hold jump button and release to make high jump.
3. Start Map Editor and press F5.

Now repeat the same steps of number 2. 

**Expected behaviour**
Everytime Freeroam starts after press Full Test your maxed out skills resets because defaultstats automatically stop so you can't do bicycle high jump or shoot while walking, for example.

**Comment by sh4z4n** [comment]
If you add the race gamemode on map settings the stats does not reset when test mode start.

**Comment by xLuxy** [comment]
Here's the line actually causing that issue: [https://github.com/multitheftauto/mtasa-resources/blob/master/%5Bgameplay%5D/defaultstats/defaultstats.lua#L30](click me). I guess that `onGamemodeMapStart` is being triggered here, removing that line might work. Question here is: Would it make sense to remove that line?

**Comment by PlatinMTA** [comment]
> Here's the line actually causing that issue: [click me](https://github.com/multitheftauto/mtasa-resources/blob/master/%5Bgameplay%5D/defaultstats/defaultstats.lua#L30)
> I guess that `onGamemodeMapStart` is being triggered here, removing that line might work. Question here is: Would it make sense to remove that line?

That doesn't look like the problem. Check out that defaultstats will give you the max ammount of biking ability, but when using the editor sh4z4n said that the maxed out skills resets, probably because realdriveby just stops.

Even if it was reseting your stats at every freeroam start, you will have 999 biking skill anyway, and you should be able to make really high jumps with that.

For your question, it seems that `onGamemodeMapStart` is kinda pointless, even if the stats reset after a gamemode change, the players would eventually spawn ig. Also i don't think that any gamemode resets your stats ever.

**Comment by xLuxy** [comment]
yeah, just noticed that, i guess i just miss-read the initial problem. gonna check again.

**Comment by Dutchman101** [comment]
Something weird is going on.. facts:

> that the maxed out skills resets, probably because defaultstats just stops.

if defaultstats stops, then it will stop applying the defined stats when it's supposed to, but not set all stats to zero. This is what currently (per this bug) happens, and the reset stats apply to more than just bikes: everything, including that in map editor you'll be aiming down the sights like this:
![image](https://user-images.githubusercontent.com/12105539/82718419-ae3ba180-9ca2-11ea-94f6-06102dfc2d05.png)

The problem is triggered by [gamemodestopper.lua:18](https://github.com/multitheftauto/mtasa-resources/blob/e51ed2fb1736c547ca0c7984d43a161d2c6872ac/%5Beditor%5D/editor_main/gamemodestopper.lua#L18), because that function calls [mapmanager_exports.lua#L193](https://github.com/multitheftauto/mtasa-resources/blob/e51ed2fb1736c547ca0c7984d43a161d2c6872ac/%5Bmanagers%5D/mapmanager/mapmanager_exports.lua#L193) and these utility functions are malfunctioning:

```lua
function isGamemode(resource)
	if type(resource) ~= "userdata" then
		return false
	end
	return (getResourceInfo(resource,"type") == "gamemode")
end

function isMap(resource)
	if type(resource) ~= "userdata" then --!w
		return false
	end
	return (getResourceInfo(resource,"type") == "map")
end
```

I believe they are malfunctioning as "defaultstats" doesn't currently have any "type" defined in its [meta.xml](https://github.com/multitheftauto/mtasa-resources/blob/master/%5Bgameplay%5D/defaultstats/meta.xml) which means it shouldn't return true, but under the circumstances of this issue (when called by editor), return true which leads to the resource mistakenly stopping.

But most of all, while I was reproducing i found that this is only the case when you do this:

> 1. Start Freeroam, then defaultstats.
> 2. Spawn any bicycle, hold jump button and release to make high jump.
> 3. Start Map Editor and press F5.

in "Host game" mode (main menu), and the problem isn't there (stats won't be nulled) when you run MTA Server.exe and perform the same steps. The reason these utility functions or something else fails only with "Host game" is yet to be investigated

The reason for stats being nulled (which like I said, normally doesn't happen when "defaultstats" stops).. my guess: `applyStatsForEveryone()` (from `onGamemodeMapStart`) is called before map editor performs the check and stops the resource, but in the midst of 'writing' stats to the game and therefore not getting the chance to complete (resulting in GTA/MTA nulling it instead, for some reason).

Edit: while writing this post, I suddenly stopped being able to reproduce the issue at all (even after restoring all resources to default).. both in "Host game" and MTA Server. So all of this can happen randomly.. what I noticed though, is that in my case it only happened when this: `WARNING: 'FREEROAM' resource is currently running.  The resource has been shut off as a precaution!` was being output even though freeroam resource was still running (it failed to stop it).

Not done debugging yet, but this all means the problem might be deeper (or perhaps in MTA)

**Comment by Dutchman101** [comment]
I edited the above comment significantly, and i want to re-iterate that I am looking for the root cause this deep because of the following:

Nowhere in the code of editor (or anywhere else) is there something that would cause all stats to be nulled (as if "defaultstats" never ran to set new values), stopping "defaultstats" normally doesn't null the stats (literally to 0 value, even worse than GTA defaults). This would only happen if "defaultstats" hasn't run at least once while the player was connected.. so yeah, something is funky, probably within MTA itself and not the resources.

**Comment by MittellBuurman** [comment]
> I edited the above comment significantly, and i want to re-iterate that I am looking for the root cause this deep because of the following:
> 
> Nowhere in the code of editor (or anywhere else) is there something that would cause all stats to be nulled (as if "defaultstats" never ran to set new values), stopping "defaultstats" normally doesn't null the stats (literally to 0 value, even worse than GTA defaults). This would only happen if "defaultstats" hasn't run at least once while the player was connected.. so yeah, something is funky, probably within MTA itself and not the resources.


There are 2 issues I found in both `defaultstats` and `play` resource.
- defaultstats uses `onGamemodeMapStart`, which I assume looks specifically for map elements. Play does not have a map defined(?). I tried using the event to trigger the call, but did not work.
- I then changed the event from `onGamemodeMapStart` to `onGamemodeStart`in `defaultstats`. This does work, but there's a caveat; on first server startup, when you restart `play`, `onGamemodeStart` does trigger, but `defaultstats` doesn't pick it up. When restarting the `defaultstats` and then restarting `play` fixes the issue.

## default mtaserver.conf and the importance of resource load order
When using a lot of resources that talk to eachother, I have found a lot of interesting things. Sometimes your function works, and sometimes it doesn't. the phrase **have you tried turning it off and on again** somehow does the trick.. but why?

The answer: resource load order.

in the default `mtaserver.conf` file, all resources are loaded in the following order:

```xml
    <resource src="admin" startup="1" protected="0" />
    <resource src="defaultstats" startup="1" protected="0" />
    <resource src="helpmanager" startup="1" protected="0" />
    <resource src="joinquit" startup="1" protected="0" />
    <resource src="mapcycler" startup="1" protected="0" />
    <resource src="mapmanager" startup="1" protected="0" />
    <resource src="parachute" startup="1" protected="0" />
    <resource src="resourcebrowser" startup="1" protected="1" default="true" />
    <resource src="resourcemanager" startup="1" protected="1" />
    <resource src="scoreboard" startup="1" protected="0" />
    <resource src="spawnmanager" startup="1" protected="0" />
    <resource src="voice" startup="1" protected="0" />
    <resource src="votemanager" startup="1" protected="0" />
    <resource src="webadmin" startup="1" protected="0" />
    <resource src="play" startup="1" protected="0" />
```

Neatly packed in alphabetical order, this list is `out of order`. `defaultstats` will load, but not function 100%

### But why?!
Resource load order is important. Since some resources are depending on other resources.
In this case, the `defaultstats` resource is loaded before `mapmanager` (actually `mapcycler`, but that's out of this scope).
    
#### mapmanager_main.lua

```lua
currentGamemode = nil
currentGamemodeMap = nil
nextGamemode = nil
nextGamemodeMap = nil

setGameType(false)
setMapName("None")

addEvent("onGamemodeStart") --> now it is registered, any resources that has an eventHandler already defined before the resource started, they are nothing but sad empty stubs.
addEvent("onGamemodeStop")
addEvent("onGamemodeMapStart") --> This is the current event used
addEvent("onGamemodeMapStop")
 ```
#### defaultstats.lua
```lua
local function applyStatsForEveryone()
	local playersTable = getElementsByType("player")

	for playerID = 1, #playersTable do
		applyStatsForPlayer(playersTable[playerID])
	end
end
addEventHandler("onResourceStart", resourceRoot, applyStatsForEveryone)
addEventHandler("onGamemodeStart", root, applyStatsForEveryone) 
                        -- > adds event before mapmanager loads, throws an exception to root (we cannot see)
```

Since the `addEventHandler` is triggered before the resource that adds the event is running. It actually throws an error. However, we cannot capture it (unless we explicitly log it) and therefore we're left with our ears scratching.

## PR will follow soon

I got the changes ready, just got to make a PR.

I'll leave this here, and tell you where you should put the `defaultstats` resource in the `.conf` file.
Putting it under `mapmanager`  is recommended.

```xml
    <resource src="admin" startup="1" protected="0" />
    <!-- <resource src="defaultstats" startup="1" protected="0" /> --> <!-- default -->
    <resource src="helpmanager" startup="1" protected="0" />
    <resource src="joinquit" startup="1" protected="0" />
    <resource src="mapcycler" startup="1" protected="0" /> <!-- mapmanager is included here -->

    <resource src="defaultstats" startup="1" protected="0" /> <!-- where it should be (at least) -->
    
    <resource src="mapmanager" startup="1" protected="0" /> <!-- where the events are added -->
    <resource src="parachute" startup="1" protected="0" />
    <resource src="resourcebrowser" startup="1" protected="1" default="true" />
    <resource src="resourcemanager" startup="1" protected="1" />
    <resource src="scoreboard" startup="1" protected="0" />
    <resource src="spawnmanager" startup="1" protected="0" />
    <resource src="voice" startup="1" protected="0" />
    <resource src="votemanager" startup="1" protected="0" />
    <resource src="webadmin" startup="1" protected="0" />

    <resource src="play" startup="1" protected="0" />
```

**Comment by jlillis** [comment]
Fixed via #471
